compiler/src/typed/translprim.re

```diff
@@ -1486,142 +1486,176 @@ let transl_prim = (env, desc) => {
   //
   // Which can pass through the rest of the compilation without issue.
   //
-  let (value, attrs) =
+  let (value, typ) =
     switch (prim) {
     | PrimitiveConstant(const) =>
-      let (value, needs_disablegc) =
+      let (value, typ, attributes) =
         switch (const) {
         // [NOTE] should be kept in sync with `runtime_heap_ptr` and friends in `compcore.re`
         | HeapBase => (
             Constant.wasmi32(string_of_int(active_memory_base())),
-            true,
+            Builtin_types.type_wasmi32,
+            disable_gc,
           )
         | HeapStart => (
             Constant.wasmi32(string_of_int(active_memory_base() + 0x10)),
-            true,
+            Builtin_types.type_wasmi32,
+            disable_gc,
           )
         | HeapTypeMetadata => (
             Constant.wasmi32(string_of_int(active_memory_base() + 0x8)),
-            true,
+            Builtin_types.type_wasmi32,
+            disable_gc,
           )
         | ElideTypeInfo => (
             Constant.bool(Grain_utils.Config.elide_type_info^),
-            false,
+            Builtin_types.type_bool,
+            [],
           )
         };
-      let attrs =
-        if (needs_disablegc) {
-          [Location.mknoloc(Typedtree.Disable_gc)];
-        } else {
-          [];
+      (Expression.constant(~loc, ~attributes, value), typ);
+    | Primitive0(p) =>
+      let attributes =
+        switch (p) {
+        | AllocateInt32
+        | AllocateInt64
+        | AllocateUint32
+        | AllocateUint64
+        | AllocateFloat32
+        | AllocateFloat64
+        | AllocateRational
+        | WasmMemorySize
+        | Unreachable => disable_gc
         };
-      (Expression.constant(~loc, ~attributes=disable_gc, value), attrs);
-    | Primitive0(
-        (
-          AllocateInt32 | AllocateInt64 | AllocateUint32 | AllocateUint64 |
-          AllocateFloat32 |
-          AllocateFloat64 |
-          AllocateRational |
-          Unreachable
-        ) as p,
-      ) => (
-        Expression.lambda(
-          ~loc,
-          ~attributes=disable_gc,
-          [],
-          Expression.prim0(~loc, p),
-        ),
-        [],
-      )
-    | Primitive1(
-        (
-          WasmUnaryI32(_) | WasmUnaryI64(_) | WasmUnaryF32(_) | WasmUnaryF64(_) |
-          WasmMemoryGrow |
-          WasmFromGrain |
-          WasmToGrain
-        ) as p,
-      ) => (
+      (
+        Expression.lambda(~loc, ~attributes, [], Expression.prim0(~loc, p)),
+        Typecore.prim0_type(p),
+      );
+    | Primitive1(BuiltinId as p) =>
```

Comment:
`BuiltinId` always means `Void` or am I missing something?

Reply:
It just means we don't generate a function for it, so you can't `List.map(builtinId, ...)`, for example. This transform is just setting up closures for primitives in the off chance you want to use them in a first-class way, but that's not an option for `BuiltinId` because it's just injecting type information from the compiler.